x/ccv/provider/keeper/punish_validator_test.go

```diff
@@ -125,8 +139,345 @@ func TestJailValidator(t *testing.T) {
+				createUndelegation([]int64{}, []time.Time{}),
+				createUndelegation([]int64{100, 100}, []time.Time{nowPlus1Hour, nowPlus1Hour}),
+				createUndelegation([]int64{800}, []time.Time{nowPlus1Hour}),
+				createUndelegation([]int64{500}, []time.Time{nowPlus1Hour}),
+			},
+			[]stakingtypes.Redelegation{},
+			int64(1),
+			sdk.NewInt(3),
+			int64(2000/3 + 1),
+		},
+		{
+			"both (mature) undelegations and redelegations",
+			// 2000 total undelegation tokens, 250 + 100 + 500 = 850 of those are from mature undelegations,
+			// so 2000 - 850 = 1150
+			[]stakingtypes.UnbondingDelegation{
+				createUndelegation([]int64{250, 250}, []time.Time{nowPlus1Hour, now}),
+				createUndelegation([]int64{}, []time.Time{}),
+				createUndelegation([]int64{100, 100}, []time.Time{now, nowPlus1Hour}),
+				createUndelegation([]int64{800}, []time.Time{nowPlus1Hour}),
+				createUndelegation([]int64{500}, []time.Time{now}),
+			},
+			// 3500 total redelegation tokens, 350 + 200 + 400 = 950 of those are from mature redelegations
+			// so 3500 - 950 = 2550
+			[]stakingtypes.Redelegation{
+				createRedelegation([]int64{}, []time.Time{}),
+				createRedelegation([]int64{1600}, []time.Time{nowPlus1Hour}),
+				createRedelegation([]int64{350, 250}, []time.Time{now, nowPlus1Hour}),
+				createRedelegation([]int64{700, 200}, []time.Time{nowPlus1Hour, now}),
+				createRedelegation([]int64{}, []time.Time{}),
+				createRedelegation([]int64{400}, []time.Time{now}),
+			},
+			int64(8391),
+			sdk.NewInt(2),
+			int64((1150+2550)/2 + 8391),
+		},
+	}
+
+	pubKey, _ := cryptocodec.FromTmPubKeyInterface(tmtypes.NewMockPV().PrivKey.PubKey())
+	validator, _ := stakingtypes.NewValidator(pubKey.Address().Bytes(), pubKey, stakingtypes.Description{})
+
+	for _, tc := range testCases {
+		gomock.InOrder(mocks.MockStakingKeeper.EXPECT().
+			SlashUnbondingDelegation(gomock.Any(), gomock.Any(), int64(0), sdk.NewDec(1)).
+			DoAndReturn(
+				func(_ sdk.Context, undelegation stakingtypes.UnbondingDelegation, _ int64, _ sdk.Dec) sdk.Int {
+					sum := sdk.NewInt(0)
+					for _, r := range undelegation.Entries {
+						if r.IsMature(ctx.BlockTime()) {
+							continue
+						}
+						sum = sum.Add(sdk.NewInt(r.InitialBalance.Int64()))
+					}
+					return sum
+				}).AnyTimes(),
+			mocks.MockStakingKeeper.EXPECT().
+				SlashRedelegation(gomock.Any(), gomock.Any(), gomock.Any(), int64(0), sdk.NewDec(1)).
+				DoAndReturn(
+					func(ctx sdk.Context, _ stakingtypes.Validator, redelegation stakingtypes.Redelegation, _ int64, _ sdk.Dec) sdk.Int {
+						sum := sdk.NewInt(0)
+						for _, r := range redelegation.Entries {
+							if r.IsMature(ctx.BlockTime()) {
+								continue
+							}
+							sum = sum.Add(sdk.NewInt(r.InitialBalance.Int64()))
+						}
+						return sum
+					}).AnyTimes(),
+		)
+
+		actualPower := providerKeeper.ComputePowerToSlash(ctx, validator,
+			tc.undelegations, tc.redelegations, tc.power, tc.powerReduction)
```

Comment:
Could you check here that afterwards, the undelegations and redelegations have not been slashed? Just to make explicit that ComputePowerToSlash does not slash on its own?

Reply:
I think it's hard to check this in the unit test here.

We could have something like:
```
tokensBeforeCall := validator.GetTokens()
delegatorSharesBeforeCall := validator.GetDelegatorShares()

actualPower := providerKeeper.ComputePowerToSlash(ctx, validator,
	tc.undelegations, tc.redelegations, tc.power, tc.powerReduction)

// safeguard check that validator remains unmodified after a call to `ComputePowerToSlash`
// `ComputePowerToSlash` only computes the power and does not modify the state of the system in any way
require.Equal(t, tokensBeforeCall, validator.GetTokens())
require.Equal(t, delegatorSharesBeforeCall, validator.GetDelegatorShares())
```
but I do not feel it actually is that useful because we actually would like to check the underlying undelegations. 
We could build an integration test for this but I'm not sure it's worth the added complexity. Additionally, this is implicitly tested in the E2E tests through the `RepresentativePowers`. If somebody were to modify the `ComputePowerToSlash` so that it actually slashes as well, then the E2E test would fail.

Reply:
Ok, I see. I thought this would be easy, but I see it's relatively complex.
Then I think it's ok to just have it in the e2e test.